src/Raven.Client/Util/StreamWithTimeout.cs

```diff
@@ -0,0 +1,141 @@
+﻿using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Raven.Client.Util
+{
+    internal class StreamWithTimeout : Stream
+    {
+        private static readonly TimeSpan DefaultWriteTimeout = TimeSpan.FromSeconds(120);
+        private static readonly TimeSpan DefaultReadTimeout = TimeSpan.FromSeconds(120);
+        
+        private readonly Stream _stream;
+        private int _writeTimeout;
+        private int _readTimeout;
+        private bool _canBaseStreamTimeoutOnWrite;
+        private bool _canBaseStreamTimeoutOnRead;
+        private CancellationTokenSource _cts;
+        public StreamWithTimeout(Stream stream, TimeSpan? writeTimeout = null, TimeSpan? readTimeout = null)
+        {
+            _stream = stream;
+            SetWriteTimeoutIfNeeded(writeTimeout);
+            SetReadTimeoutIfNeeded(readTimeout);
+        }
+
+        private void SetReadTimeoutIfNeeded(TimeSpan? readTimeout)
+        {
+            _canBaseStreamTimeoutOnRead = _stream.CanTimeout && _stream.ReadTimeout < int.MaxValue;
+
+            if (_canBaseStreamTimeoutOnRead)
+                _readTimeout = (int?)readTimeout?.TotalMilliseconds ?? _stream.ReadTimeout;
+            else
+                _readTimeout = (int)(readTimeout ?? DefaultReadTimeout).TotalMilliseconds;
+        }
+
+        private void SetWriteTimeoutIfNeeded(TimeSpan? writeTimeout)
+        {
+            _canBaseStreamTimeoutOnWrite = _stream.CanTimeout && _stream.WriteTimeout < int.MaxValue;
+
+            if (_canBaseStreamTimeoutOnWrite)
+                _writeTimeout = (int?)writeTimeout?.TotalMilliseconds ?? _stream.WriteTimeout;
```

Comment:
Shouldn't we set the `WriteTimeout` here? If there is a smaller timeout already defined?

Reply:
You right, we need to set the inner stream timeouts as well.

About the smaller value, I think that if we specify a timeout is should take precedence over the existing one,